data/tutorials/ds_05_seq.md

```diff
@@ -0,0 +1,317 @@
+```ocaml
+# facts |> take 100 |> List.of_seq;;
+- : int list =
+[2; 3; 5; 7; 11; 13; 17; 19; 23; 29; 31; 37; 41; 43; 47; 53; 59; 61; 67; 71;
+ 73; 79; 83; 89; 97; 101; 103; 107; 109; 113; 127; 131; 137; 139; 149; 151;
+ 157; 163; 167; 173; 179; 181; 191; 193; 197; 199; 211; 223; 227; 229; 233;
+ 239; 241; 251; 257; 263; 269; 271; 277; 281; 283; 293; 307; 311; 313; 317;
+ 331; 337; 347; 349; 353; 359; 367; 373; 379; 383; 389; 397; 401; 409; 419;
+ 421; 431; 433; 439; 443; 449; 457; 461; 463; 467; 479; 487; 491; 499; 503;
+ 509; 521; 523]
+```
+
+The function `sieve` is recursive, in OCaml and common senses: it is defined
+using the `rec` keyword and calls itself. However, some call that kind of
+function [_corecursive_](https://en.wikipedia.org/wiki/Corecursion). This word
+is used to emphasize that, by design, it does not terminate. Strictly speaking,
+the sieve of Eratosthenes is not an algorithm either since it does not
+terminate. This implementation behaves the same.
+
+## Unfolding Sequences
+
+Standard higher-order iteration functions are available on sequences. For
+instance:
+* `Seq.iter`
+* `Seq.map`
+* `Seq.fold_left`
+
+All those are also available for `Array`, `List` and `Set` and behave
+essentially the same. Observe that there is no `fold_right` function. Since
+OCaml 4.11 there is something which isn't (yet) available on other types:
+`unfold`. Here is how it is implemented:
+```ocaml
+let rec unfold f seq () = match f seq with
+  | None -> Nil
+  | Some (x, seq) -> Cons (x, unfold f seq)
+```
+And here is its type:
+```ocaml
+val unfold : ('a -> ('b * 'a) option) -> 'a -> 'b Seq.t = <fun>
+```
+Unlike previously mentioned iterators `Seq.unfold` does not have a sequence
+parameter, but a sequence result. `unfold` provides a general means to build
+sequences. For instance, `Seq.ints` can be implemented using `Seq.unfold` in a
+fairly compact way:
+```ocaml
+let ints = Seq.unfold (fun n -> Some (n, n + 1));;
+```
+
+As a fun fact, one should observe `map` over sequences can be implemented using
+`Seq.unfold`. Here is how to write it:
+```ocaml
+# let map f = Seq.unfold (fun s -> s |> Seq.uncons |> Option.map (fun (x, y) -> (f x, y)));;
+val map : ('a -> 'b) -> 'a Seq.t -> 'b Seq.t = <fun>
+```
+Here is a quick check:
+```ocaml
+# Seq.ints 0 |> map (fun x -> x * x) |> Seq.take 10 |> List.of_seq;;
+- : int list = [0; 1; 4; 9; 16; 25; 36; 49; 64; 81]
+```
+The function `Seq.uncons` returns the head and tail of a sequence if it is not empty or `None` otherwise.
+
+Using this function:
+```ocaml
+let input_line_opt chan =
+  try Some (input_line chan, chan)
+  with End_of_file -> close_in chan; None
+```
+It is possible to read a file using `Seq.unfold`:
+```ocaml
+"README.md" |> open_in |> Seq.unfold input_line_opt |> Seq.iter print_endline
+```
+
+Although this can be an appealing style, bear in mind it does not prevent from
+taking care of open files. While the code above is fine, this one no longer is:
```

Comment:
```suggestion
Although this can be an appealing style, bear in mind that it does not prevent
taking care of open files. While the code above is fine, this one no longer is:
```